### PR TITLE
new definitions for zipWithX

### PR DESCRIPTION
**Pull request by jippiedoe**
## Description
See #452
 
## How has this been tested?
Only really tested zipWith3, zipWith4 and izipWith4, and they both fuse great in my minimal example. There are examples where it still doesn't fuse optimally, but those are not part of this story :)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by jippiedoe** [comment]
iZipWith functions suffer from the same problem, but this also goes for the 2-tuple case so it is less easily fixed. To properly fuse these, the ast should be modified (probably by making iZipWith the standard, and having zipWith be a 'special case' where the index is not used).
Once iZipWith is fusable, it is trivial to also fuse iZipWith3 and above.

I don't know how often these functions are even used.

**Comment by jippiedoe** [comment]
Maybe I was too optimistic..
This fuses, with the new zipWith3 definition:
`simpleTest xs = zipWith3 (\a b c -> a * b * c) ys ys zs`
`  where ys = map (+1) xs`
`            zs = map (+2) xs`

But this doesn't fuse, because the shared terms aren't directly zipped together:
`simpleTest xs = zipWith3 (\a b c -> a * b * c) ys zs ys`
`  where ys = map (+1) xs`
`            zs = map (+2) xs`

Regardless, it seems like the new definition is simpler and at least generates an AST on which fusion is possible, even if currently it isn't detected yet. And I think the new definition does allow strictly more fusion than the current one, at no cost.

File: src/Data/Array/Accelerate/Prelude.hs [hunk 1]
```diff
@@ -291,13 +277,8 @@ zipWith9
...
     -> Acc (Array sh i)
     -> Acc (Array sh j)
-zipWith9 f as bs cs ds es fs gs hs is
-  = generate (shape as `intersect` shape bs `intersect` shape cs
-                       `intersect` shape ds `intersect` shape es
-                       `intersect` shape fs `intersect` shape gs
-                       `intersect` shape hs `intersect` shape is)
-             (\ix -> f (as ! ix) (bs ! ix) (cs ! ix) (ds ! ix) (es ! ix) (fs ! ix) (gs ! ix) (hs ! ix) (is ! ix))
-
+zipWith9 fn as bs cs ds es fs gs hs is
```
Comment (ivogabe) [inline src/Data/Array/Accelerate/Prelude.hs]: I think you can eta reduce this to:

```hs
zipWith9 fn as bs
  = zipWith8 (\(unlift -> (a, b)) -> fn a b) (zip as bs)
```

Comment (jippiedoe) [reply src/Data/Array/Accelerate/Prelude.hs]: How is this? :)

Comment (ivogabe) [reply src/Data/Array/Accelerate/Prelude.hs]: Great idea, now it’s even less code

**Comment by ivogabe** [comment]
You could use patterns instead of unlifting (eg `T2`). Besides that, looks good and can be merged in my opinion.

**Comment by jippiedoe** [comment]
Changed to use these patterns, I wanted to change the rest of the prelude to use "T2" and "::." too but there were just too many. If someone is bored one day, they could replace all 90 remaining instances of 'lift' with the appropriate pattern at some point.

@tmcdonell I think this is ready to merge

**Pull request merged**